### PR TITLE
Refactor YamuxHandler.SendBuffer

### PR DESCRIPTION
**Pull request by Nashatyrev**
* Introduce ByteBufQueue
* Add ByteBufQueue tests
* Writing exec path is always through fill/drain buffer
* Adopt/fix existing tests
* Add new test checking correct handling of negative sendWindowSize

### CODE REVIEW

**Review by StefanBratanov — APPROVED** [review]
LGTM

File: libp2p/src/main/kotlin/io/libp2p/etc/util/netty/ByteBufQueue.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        val partialBufferSlice =
+            when {
+                data.isEmpty() -> null
+                size == maxLength -> null
+                else -> data.first()
+            }
+                ?.let { buf ->
+                    val remainingBytes = maxLength - size
+                    buf.readSlice(remainingBytes).retain()
```
Comment (StefanBratanov) [inline libp2p/src/main/kotlin/io/libp2p/etc/util/netty/ByteBufQueue.kt:32]: I think can use `readRetainedSlice()` instead. At least docs state it could be better.
